web/modules/custom/rdf_ff/src/SchemaFieldValidator.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Drupal\rdf_ff;
+
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\rdf_entity\Database\Driver\sparql\Connection;
+use Drupal\Core\Entity\EntityTypeManager;
+use Drupal\rdf_entity\Entity\Query\Sparql\SparqlArg;
+use Drupal\rdf_entity\RdfEntityMappingInterface;
+use Drupal\rdf_entity\RdfFieldHandler;
+
+/**
+ * A service that validates that fields are defined in a schema.
+ */
+class SchemaFieldValidator implements SchemaFieldValidatorInterface {
+
+  /**
+   * The database connection.
+   *
+   * @var \Drupal\rdf_entity\Database\Driver\sparql\Connection
+   */
+  protected $sparqlEndpoint;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManager
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The rdf field handler service.
+   *
+   * @var \Drupal\rdf_entity\RdfFieldHandler
+   */
+  protected $fieldHanlder;
+
+  /**
+   * Constructs a new SchemaFieldValidator object.
+   *
+   * @param \Drupal\rdf_entity\Database\Driver\sparql\Connection $sparql_endpoint
+   *   The database connection.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\rdf_entity\RdfFieldHandler $field_hanlder
+   *   The rdf field handler service.
+   */
+  public function __construct(Connection $sparql_endpoint, EntityTypeManagerInterface $entity_type_manager, RdfFieldHandler $field_hanlder) {
+    $this->sparqlEndpoint = $sparql_endpoint;
+    $this->entityTypeManager = $entity_type_manager;
+    $this->fieldHanlder = $field_hanlder;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Exception
+   *   Thrown if the entity does not have mapped properties.
+   */
+  public function isDefinedInSchema($entity_type_id, $bundle, $field_name, $column_name = '') {
+    $mapping = $this->getEntityMapping($entity_type_id, $bundle);
+    if (empty($mapping) || empty($properties = $mapping->getThirdPartySettings('rdf_ff'))) {
+      throw new \Exception("The entity does not appear to have mapped properties.");
+    }
+
+    $predicate = $this->fieldHanlder->getFieldPredicates($entity_type_id, $field_name, $column_name);
+    $rdf_type = $mapping->getRdfType();
+    $query = $this->getQuery($properties['graph'], $properties['property_predicates'], $rdf_type, $predicate[$bundle]);
+
+    return $this->sparqlEndpoint->query($query)->isTrue();
+  }
+
+  /**
+   * Retrieves an RdfEntityMapping entity.
+   *
+   * @param string $entity_type_id
+   *   The entity type id.
+   * @param string $bundle
+   *   The entity type id.
+   *
+   * @return \Drupal\rdf_entity\RdfEntityMappingInterface|null
+   *   The rdf entity mapping or null, if none is found.
+   */
+  protected function getEntityMapping($entity_type_id, $bundle): ?RdfEntityMappingInterface {
```

Comment:
I was taught by Pieter that we should not use static methods in OOP as it is hard to mock the classes (if I am not mistaking). this is there because I am using the passed entity type manager instead of the global one.

Reply:
Hm, I heard this too from him but I have to disagree. The static call, calls also the entity type manager. In both cases you'll need to mock the EntityTypeManager. The static call will work from unit test because that calls the mocked manager. Also, I found no documentation related to this, no recommendation, no "best practices". I think this deserves, eventually to be well documented somewhere and for this reason I've opened an upstream issue: https://www.drupal.org/project/drupal/issues/2945539. But till I don't have an "official" point of view from the community I'm keeping my point of view, that the static call shorthand is correct used anywhere.

Reply:
First of all, from an OO perspective we should always use injected dependencies and not rely on hardcoded implementations. In OOP a class is intented to be a reusable object, and any dependencies it has should be declared through interfaces. This makes it architecture agnostic. This OOP pattern is called [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
In fact the dependency injection framework we use in Drupal is even from another architecture, we are using the [Symfony service container](https://symfony.com/doc/3.4/service_container.html) component. Our services should be exchangeable with services written for Symfony with little to no adaptations. If we call `RdfEntityMapping::load()` in our code the class suddenly becomes hard coupled to the Drupal 8 implementation of entities, and is not fully reusable. What is particularly bad about it is that it hard couples to an _implementation_ (`\Drupal\rdf_entity\Entity\RdfEntityMapping` in full). If it was a static method defined on an interface it wouldn't even be _that_ bad, but by specifying the implementation itself we can no longer override it, we are married to that specific implementation.

Secondly, the one and only reason that we have `EntityInterface::load()` and `EntityInterface::loadMultiple()` is because this made the task of porting core and contrib from Drupal 7 to Drupal 8 vastly easier. They are intended to be used as direct replacements for `entity_load()` and `entity_load_multiple()` (and their entity specific cousins like `node_load()` and friends) _in procedural code_.  It was never intended to be used in OO classes.

At the time this was introduced (ref. https://www.drupal.org/node/2190313) we were in the middle of porting legacy Drupal 7 code and were faced with replacing tens of thousands of instances of `entity_load()` with `\Drupal::entityManager()->getStorage($entity_type_id)->load($entity_id)`. We were concerned about the bad DX of this pattern and people were actively complaining that D8 code was too hard to write.

Thirdly, this makes it unnecessarily difficult to unit test. In a proper unit test only the class itself (the "System under test") is loaded, and all other dependencies are represented by controlled mocks. If we call `RdfEntityMapping::loadByName()` then suddenly this class gets loaded and the method executed, which is something we don't want in a controlled unit test. To make it worse, in turn this code will load the entire `\Drupal` class statically. This continues until it finally throws a `ContainerNotInitializedException` because in a unit test the container doesn't exist. I've seen some unit tests that work around this creatively, but if we simply follow the standard OOP pattern then we can simply mock the entity storage properly.

Reply:
@pfrenssen could we have this opinion copied in the upstream ticket? https://www.drupal.org/project/drupal/issues/2945539. I want to answer there because, finally, we want this written somewhere as being (or not) a pattern that we should follow.

Reply:
I also posted a reply in the upstream issue: https://www.drupal.org/project/drupal/issues/2945539#comment-12489703

Reply:
OK, thanks, I'll reply there. DX is dead :)

Reply:
I'm keeping here my point. I specially created `::loadByName()` as a shorthand, to avoid crating such useless method (in fact is mocking EntityInterface::load()). But let's go with the actual code, I'm OK